### PR TITLE
Add bleu

### PR DESCRIPTION
**Pull request by coder1729**
Added BLEU from AllenNLP instead of what @vince62s suggested as there tensors can be used directly and minibatches with validation would work. In @vince62s 's reference of corpus_bleu, entire validation set would be required. Also this already excluded the pad tokens.

https://github.com/OpenNMT/OpenNMT-py/issues/1158

### CODE REVIEW

**Comment by vince62s** [comment]
Thanks for the contribution, however you will need to change the logic.
As a first PR you just need to add Bleu as an extra validation metric (look at how accuracy and ppl are done in the validation process). You don't need to change the loss related functions.
Maybe later on we can change the loss fucntion itself to implement other things but this is not the point at this stage.

**Comment by coder1729** [comment]
@vince62s so do you want all the members of the bleu class to be included in the Statistics class and updations to precision_matches etc as a part of the batch_stats update?

**Comment by vince62s** [comment]
no, the class in its own file is fine.
I am just saying that Bleu is another metric at the same level as PPL or ACC
that's it.

**Comment by coder1729** [comment]
@vince62s  please see how it looks now. Thanks !

File: onmt/trainer.py [hunk 1]
```diff
@@ -195,7 +194,8 @@ def train(self, train_iter, valid_iter, train_steps, valid_steps):
                                 logger.info('GpuRank %d: report stat step %d'
                                             % (self.gpu_rank, step))
                             self._report_step(self.optim.learning_rate,
-                                              step, valid_stats=valid_stats)
+                                              step, valid_stats=valid_stats,
+                                              )
```
Comment (vince62s) [inline onmt/trainer.py]: fix this please.

File: onmt/utils/loss.py [hunk 2]
```diff
@@ -86,7 +87,8 @@ def __init__(self, criterion, generator):
     def padding_idx(self):
         return self.criterion.ignore_index
 
-    def _make_shard_state(self, batch, output, range_, attns=None):
+    def _make_shard_state(self, batch, output, range_, attns=None,
+                          ):
         """
```
Comment (vince62s) [inline onmt/utils/loss.py]: fix this please.

File: onmt/utils/loss.py [hunk 3]
```diff
@@ -128,7 +130,8 @@ def monolithic_compute_loss(self, batch, output, attns):
             :obj:`onmt.utils.Statistics`: loss statistics
         """
         range_ = (0, batch.tgt.size(0))
-        shard_state = self._make_shard_state(batch, output, range_, attns)
+        shard_state = self._make_shard_state(batch, output, range_, attns,
+                                             )
         _, batch_stats = self._compute_loss(batch, **shard_state)
```
Comment (vince62s) [inline onmt/utils/loss.py]: fix this please

File: onmt/utils/loss.py [hunk 4]
```diff
@@ -172,21 +175,136 @@ def sharded_compute_loss(self, batch, output, attns,
...
         Returns:
             :obj:`onmt.utils.Statistics` : statistics for this batch.
         """
         pred = scores.max(1)[1]
+        precision_matches = Counter()
+        precision_totals = Counter()
+        prediction_lengths = 0
+        reference_lengths = 0
+
+        if batch_sz is not None:
```
Comment (vince62s) [inline onmt/utils/loss.py]: don't really understand why we need the batch_size for.

File: onmt/utils/loss.py [hunk 5]
```diff
@@ -246,8 +364,7 @@ def _compute_loss(self, batch, output, target):
         gtruth = target.view(-1)
 
         loss = self.criterion(scores, gtruth)
-        stats = self._stats(loss.clone(), scores, gtruth)
-
+        stats = self._stats(loss.clone(), scores, gtruth, batch.tgt.size(1))
```
Comment (vince62s) [inline onmt/utils/loss.py]: why do we need this? (see my comment above. we should do without.

File: onmt/utils/report_manager.py [hunk 6]
```diff
@@ -84,17 +84,20 @@ def _report_training(self, *args, **kwargs):
         """ To be overridden """
         raise NotImplementedError()
 
-    def report_step(self, lr, step, train_stats=None, valid_stats=None):
+    def report_step(self, lr, step, train_stats=None, valid_stats=None,
+                    ):
         """
```
Comment (vince62s) [inline onmt/utils/report_manager.py]: fix this

File: onmt/utils/report_manager.py [hunk 7]
```diff
@@ -84,17 +84,20 @@ def _report_training(self, *args, **kwargs):
...
         Args:
             train_stats(Statistics): training stats
             valid_stats(Statistics): validation stats
             lr(float): current learning rate
+            bleu(Bleu): Bleu object
         """
         self._report_step(
-            lr, step, train_stats=train_stats, valid_stats=valid_stats)
+            lr, step, train_stats=train_stats, valid_stats=valid_stats,
+            )
```
Comment (vince62s) [inline onmt/utils/report_manager.py]: fix this

File: onmt/utils/report_manager.py [hunk 8]
```diff
@@ -84,17 +84,20 @@ def _report_training(self, *args, **kwargs):
...
+                    ):
         """
         Report stats of a step
 
         Args:
             train_stats(Statistics): training stats
             valid_stats(Statistics): validation stats
             lr(float): current learning rate
+            bleu(Bleu): Bleu object
         """
```
Comment (vince62s) [inline onmt/utils/report_manager.py]: remove this

File: onmt/utils/loss.py (hunk 4 above)
Comment (coder1729) [reply onmt/utils/loss.py]: _get_modified_precision_counts assumes the shape as (batch,seq_length) before this it is not that.

Comment (vince62s) [reply onmt/utils/loss.py]: what I meant is that I don't really see why you need to do the calculation here in the loss.py module.
it should only happen in the statistics.py
as I said before, you need to keep all the Bleu classes / module in its own bleupy so that it can be reused elsewhere, but so far we just need Bleu as a stat, not a trigger for training.

**Comment by coder1729** [comment]
@vince62s  to remove batch size a small refactor is required. Please have a look, also moved to bleupy

File: onmt/modules/copy_generator.py [hunk 9]
```diff
@@ -196,9 +197,13 @@ def _compute_loss(self, batch, output, target, copy_attn, align):
...
         target_data[correct_mask] += offset_align
-
         # Compute sum of perplexities for stats
-        stats = self._stats(loss.sum().clone(), scores_data, target_data)
+
+        stats = self._stats(loss.sum().clone(),
+                            scores_data.view(
+                                -1, batch.batch_size, scores_data.size(1)),
+                            target_data.view(
+                                -1, batch.batch_size))
```
Comment (vince62s) [inline onmt/modules/copy_generator.py]: why are you changing the copy_generator ? should not be necessary in this PR

File: onmt/utils/loss.py [hunk 10]
```diff
@@ -10,6 +10,7 @@
 import onmt
 from onmt.modules.sparse_losses import SparsemaxLoss
 from onmt.modules.sparse_activations import LogSparsemax
+from onmt.utils.bleu import bleu_calculations
```
Comment (vince62s) [inline onmt/utils/loss.py]: loos.py should not change

File: onmt/utils/loss.py [hunk 11]
```diff
@@ -178,15 +179,27 @@ def _stats(self, loss, scores, target):
             loss (:obj:`FloatTensor`): the loss computed by the loss criterion.
             scores (:obj:`FloatTensor`): a score for each possible output
             target (:obj:`FloatTensor`): true targets
+            batch_sz (:'int'): batch size
```
Comment (vince62s) [inline onmt/utils/loss.py]: remove

File: onmt/utils/loss.py [hunk 12]
```diff
@@ -178,15 +179,27 @@ def _stats(self, loss, scores, target):
...
-        num_correct = pred.eq(target).masked_select(non_padding).sum().item()
+        pred = scores.max(2)[1]
+        precision_matches, precision_totals, \
+            prediction_lengths, reference_lengths = \
+            bleu_calculations(
+                    pred,
+                    target,
+                    4,
+                    {self.padding_idx})
+
```
Comment (vince62s) [inline onmt/utils/loss.py]: remove, again loss.py should not change

File: onmt/utils/loss.py [hunk 13]
```diff
@@ -178,15 +179,27 @@ def _stats(self, loss, scores, target):
...
+        non_padding = target.view(-1).ne(self.padding_idx)
+        num_correct = pred.view(-1).eq(target.view(-1)).\
+            masked_select(non_padding).sum().item()
         num_non_padding = non_padding.sum().item()
-        return onmt.utils.Statistics(loss.item(), num_non_padding, num_correct)
+        return onmt.utils.Statistics(loss.item(), num_non_padding, num_correct,
+                                     precision_matches, precision_totals,
+                                     prediction_lengths, reference_lengths)
```
Comment (vince62s) [inline onmt/utils/loss.py]: revert back to original

File: onmt/utils/loss.py [hunk 14]
```diff
@@ -246,8 +259,9 @@ def _compute_loss(self, batch, output, target):
         gtruth = target.view(-1)
 
         loss = self.criterion(scores, gtruth)
-        stats = self._stats(loss.clone(), scores, gtruth)
 
+        scores = scores.view(-1, batch.tgt.size(1), scores.size(1))
+        stats = self._stats(loss.clone(), scores, target)
         return loss, stats
```
Comment (vince62s) [inline onmt/utils/loss.py]: revert

File: onmt/utils/report_manager.py [hunk 15]
```diff
@@ -136,7 +136,8 @@ def _report_training(self, step, num_steps, learning_rate,
 
         return report_stats
 
-    def _report_step(self, lr, step, train_stats=None, valid_stats=None):
+    def _report_step(self, lr, step, train_stats=None, valid_stats=None,
+                     ):
         """
```
Comment (vince62s) [inline onmt/utils/report_manager.py]: fix this

File: onmt/utils/loss.py (hunk 12 above)
Comment (coder1729) [reply onmt/utils/loss.py]: I am not able to not change loss.py at all and achieve this.

**Comment by mhn226** [comment]
Hi. It's really nice to have this feature. I tried this and I have a silly question about it. Applying this feature, I witnessed very small bleu score on the dev set during training. However, when I used the same models to translate the same dev set, and calculated the bleu score of the results, I found that the bleu scores were much higher (and more reasonable), even with the beam size = 1.
I am trying to understand why it appears to be different in these cases, but I don't have the answer for now. 
I wonder if I'm wrong somewhere or do you guys see the same problem? Do you have any idea about this?

**Comment by mhn226** [comment]
Hi again,
I tried to investigate this problem and I found that the BLEU score varied greatly when the valid_batch_size changed significantly. In details, when increasing the valid_batch_size, the number of predicted tokens seemed to increase, and consequently, the precision_totals also increased (true for every kind of n-grams). Whereas, the precision_matches didn't seem to dependent on the valid_batch_size. The count clip seemed to work fine here! 
So to summary, when the valid_batch_size increases, for every kind of n-grams, the precision_totals increases significantly, while the precision_matches, consequently, this obviously decreases the bleu score.
I prefer to set valid_batch_size = 1 in this case.

**Comment by BramVanroy** [comment]
Any update on this one? Would be great to have this. Considering that it is implemented in the TF version, some inspiration can be found there.

Preferably the BLEU implementation that is used should be identical between the TF and PT version of onmt (!)
